File: JavaLibrary/src/test/java/com/flat/simplex/parser/logic/line/call/CallArrayTest.java
package com.flat.simplex.parser.logic.line.call;

import com.flat.simplex.parser.logic.Context;
import com.flat.simplex.parser.logic.block.BlockIf;
import com.flat.simplex.parser.logic.error.Error;
import com.flat.simplex.support.TokenChain;
import org.junit.jupiter.api.Test;

import static com.flat.simplex.support.ContextSupport.assertErrors;
import static com.flat.simplex.support.TokenChain.readChain;
import static org.junit.jupiter.api.Assertions.*;

class CallArrayTest {

    @Test
    public void loadArray() {
        TokenChain chain = readChain("[1]");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallArray call = new CallArray(blockIf, chain.get());
        call.load();

        assertEquals(1, call.getMembers().size(), "Invalid parameters count");

        assertErrors(context);
    }

    @Test
    public void loadMultipleMemberArray() {
        TokenChain chain = readChain("[1, 2, 3]");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallArray call = new CallArray(blockIf, chain.get());
        call.load();

        assertEquals(3, call.getMembers().size(), "Invalid parameters count");

        assertErrors(context);
    }

    @Test
    public void loadMap() {
        TokenChain chain = readChain("[a : b]");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallArray call = new CallArray(blockIf, chain.get());
        call.load();

        assertEquals(1, call.getMembers().size(), "Invalid parameters count");

        assertErrors(context);
    }

    @Test
    public void loadMultipleMap() {
        TokenChain chain = readChain("[a : b, c : d]");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallArray call = new CallArray(blockIf, chain.get());
        call.load();

        assertEquals(2, call.getMembers().size(), "Invalid parameters count");

        assertErrors(context);
    }

    @Test
    public void loadGrid() {
        TokenChain chain = readChain("[[1]]");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallArray call = new CallArray(blockIf, chain.get());
        call.load();

        assertEquals(1, call.getMembers().size(), "Invalid parameters count");

        assertErrors(context);
    }

    @Test
    public void loadMultipleGrid() {
        TokenChain chain = readChain("[[1, 2], [3, 4]]");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallArray call = new CallArray(blockIf, chain.get());
        call.load();

        assertEquals(2, call.getMembers().size(), "Invalid parameters count");

        assertErrors(context);
    }

    @Test
    public void loadGridExpression() {
        TokenChain chain = readChain("[[1 + 5, 2], [3, 4]]");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallArray call = new CallArray(blockIf, chain.get());
        call.load();

        assertEquals(2, call.getMembers().size(), "Invalid parameters count");

        assertErrors(context);
    }

    @Test
    public void loadTernaryOnKeyMap() {
        TokenChain chain = readChain("[a ? b : c : d, e : f]");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallArray call = new CallArray(blockIf, chain.get());
        call.load();

        assertEquals(2, call.getMembers().size(), "Invalid parameters count");

        assertErrors(context);
    }

    @Test
    public void loadEmptyArray() {
        TokenChain chain = readChain("[]");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallArray call = new CallArray(blockIf, chain.get());
        call.load();

        assertEquals(0, call.getMembers().size(), "Invalid parameters count");

        assertErrors(context);
    }

    @Test
    public void loadEmptyMap() {
        TokenChain chain = readChain("[:]");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallArray call = new CallArray(blockIf, chain.get());
        call.load();

        assertEquals(0, call.getMembers().size(), "Invalid parameters count");

        assertErrors(context);
    }

    @Test
    public void loadEmptyGrid() {
        TokenChain chain = readChain("[[]]");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallArray call = new CallArray(blockIf, chain.get());
        call.load();

        assertEquals(0, call.getMembers().size(), "Invalid parameters count");

        assertErrors(context);
    }

    @Test
    public void loadMixTypesArrayGrid() {
        TokenChain chain = readChain("[1, [3, 4]]");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallArray call = new CallArray(blockIf, chain.get());
        call.load();

        assertEquals(2, call.getMembers().size(), "Invalid parameters count");

        assertErrors(context, Error.arrayMixingTypes);
    }

    @Test
    public void loadMixTypesArrayMap() {
        TokenChain chain = readChain("[1, a : b]");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallArray call = new CallArray(blockIf, chain.get());
        call.load();

        assertEquals(2, call.getMembers().size(), "Invalid parameters count");

        assertErrors(context, Error.arrayMixingTypes);
    }

    @Test
    public void loadMixTypesGridMap() {
        TokenChain chain = readChain("[[a], a : b]");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallArray call = new CallArray(blockIf, chain.get());
        call.load();

        assertEquals(2, call.getMembers().size(), "Invalid parameters count");

        assertErrors(context, Error.arrayMixingTypes);
    }

    @Test
    public void loadDoubleComma_Fail() {
        TokenChain chain = readChain("[a,,b]");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallArray call = new CallArray(blockIf, chain.get());
        call.load();

        assertEquals(2, call.getMembers().size(), "Invalid parameters count");

        assertErrors(context, Error.unexpectedToken);
    }

    @Test
    public void loadSingleComma_Fail() {
        TokenChain chain = readChain("[,]");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallArray call = new CallArray(blockIf, chain.get());
        call.load();

        assertEquals(0, call.getMembers().size(), "Invalid parameters count");

        assertErrors(context, Error.unexpectedToken);
    }

    @Test
    public void loadEndComma_Fail() {
        TokenChain chain = readChain("[a,]");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallArray call = new CallArray(blockIf, chain.get());
        call.load();

        assertEquals(1, call.getMembers().size(), "Invalid parameters count");

        assertErrors(context, Error.unexpectedEndOfTokens);
    }

    @Test
    public void loadCommaBetweenTernary_Fail() {
        TokenChain chain = readChain("[a ? b , : c]");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallArray call = new CallArray(blockIf, chain.get());
        call.load();

        assertEquals(2, call.getMembers().size(), "Invalid parameters count");

        assertErrors(context, Error.lineTernaryIncomplete, Error.lineTernaryIncomplete);
    }

    @Test
    public void loadMapArray_Fail() {
        TokenChain chain = readChain("[a : [1]]");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallArray call = new CallArray(blockIf, chain.get());
        call.load();

        assertEquals(1, call.getMembers().size(), "Invalid parameters count");

        assertErrors(context, Error.arrayContainer);
    }

    @Test
    public void loadMapKeyArray_Fail() {
        TokenChain chain = readChain("[[a] : 1]");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallArray call = new CallArray(blockIf, chain.get());
        call.load();

        assertEquals(1, call.getMembers().size(), "Invalid parameters count");

        assertErrors(context, Error.arrayContainer);
    }

    @Test
    public void loadGridArray_Fail() {
        TokenChain chain = readChain("[[[a]]]");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallArray call = new CallArray(blockIf, chain.get());
        call.load();

        assertEquals(1, call.getMembers().size(), "Invalid parameters count");

        assertErrors(context, Error.arrayContainer);
    }

    @Test
    public void loadMissingCloser_Fail() {
        TokenChain chain = readChain("[1");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallArray call = new CallArray(blockIf, chain.get());
        call.load();

        assertEquals(1, call.getMembers().size(), "Invalid parameters count");

        assertErrors(context, Error.missingCloser);
    }

    @Test
    public void loadMissingInnerCloser_Fail() {
        TokenChain chain = readChain("[[1");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallArray call = new CallArray(blockIf, chain.get());
        call.load();

        assertEquals(1, call.getMembers().size(), "Invalid parameters count");

        assertErrors(context, Error.missingCloser, Error.missingCloser);
    }

    @Test
    public void loadInnerComma_Fail() {
        TokenChain chain = readChain("[[1,]]");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallArray call = new CallArray(blockIf, chain.get());
        call.load();

        assertEquals(1, call.getMembers().size(), "Invalid parameters count");

        assertErrors(context, Error.unexpectedEndOfTokens);
    }

    @Test
    public void loadInnerDoubleComma_Fail() {
        TokenChain chain = readChain("[[a,,b]]");
        Context context = new Context();
        BlockIf blockIf = getBlock(context);
        CallArray call = new CallArray(blockIf, chain.get());
        call.load();

        assertEquals(1, call.getMembers().size(), "Invalid parameters count");

        assertErrors(context, Error.unexpectedToken);
    }

    private BlockIf getBlock(Context context) {
        TokenChain chain = readChain("if(true);");

        return new BlockIf(context, null, chain.get(), null);

    }
}

File: SimplexApp/simplex/parser/logic/line/call/CallArray.test.js
const TokenChain = require("simplex/support/TokenChain.js");
const Context = require("simplex/parser/logic/Context.js");
const ContextSupport = require("simplex/support/ContextSupport.js");
const Error = require("simplex/parser/logic/error/Error.js");
const CallArray = require("simplex/parser/logic/line/call/CallArray.js");

const readChain = TokenChain.readChain;
const assertErrors = ContextSupport.assertErrors;

function assertEquals(a, b, message) {
    expect(a, message).toEqual(b);
}

test('loadArray', () => {
    let chain = readChain('[1]');
    let context = new Context(chain.get());
    let call = new CallArray(context, chain.get());
    call.load();
    assertEquals(1, call.getMembers().length, 'Invalid parameters count');
    assertErrors(context);
});
test('loadMultipleMemberArray', () => {
    let chain = readChain('[1, 2, 3]');
    let context = new Context(chain.get());
    let call = new CallArray(context, chain.get());
    call.load();
    assertEquals(3, call.getMembers().length, 'Invalid parameters count');
    assertErrors(context);
});
test('loadMap', () => {
    let chain = readChain('[a : b]');
    let context = new Context(chain.get());
    let call = new CallArray(context, chain.get());
    call.load();
    assertEquals(1, call.getMembers().length, 'Invalid parameters count');
    assertErrors(context);
});
test('loadMultipleMap', () => {
    let chain = readChain('[a : b, c : d]');
    let context = new Context(chain.get());
    let call = new CallArray(context, chain.get());
    call.load();
    assertEquals(2, call.getMembers().length, 'Invalid parameters count');
    assertErrors(context);
});
test('loadGrid', () => {
    let chain = readChain('[[1]]');
    let context = new Context(chain.get());
    let call = new CallArray(context, chain.get());
    call.load();
    assertEquals(1, call.getMembers().length, 'Invalid parameters count');
    assertErrors(context);
});
test('loadMultipleGrid', () => {
    let chain = readChain('[[1, 2], [3, 4]]');
    let context = new Context(chain.get());
    let call = new CallArray(context, chain.get());
    call.load();
    assertEquals(2, call.getMembers().length, 'Invalid parameters count');
    assertErrors(context);
});
test('loadGridExpression', () => {
    let chain = readChain('[[1 + 5, 2], [3, 4]]');
    let context = new Context(chain.get());
    let call = new CallArray(context, chain.get());
    call.load();
    assertEquals(2, call.getMembers().length, 'Invalid parameters count');
    assertErrors(context);
});
test('loadTernaryOnKeyMap', () => {
    let chain = readChain('[a ? b : c : d, e : f]');
    let context = new Context(chain.get());
    let call = new CallArray(context, chain.get());
    call.load();
    assertEquals(2, call.getMembers().length, 'Invalid parameters count');
    assertErrors(context);
});
test('loadEmptyArray', () => {
    let chain = readChain('[]');
    let context = new Context(chain.get());
    let call = new CallArray(context, chain.get());
    call.load();
    assertEquals(0, call.getMembers().length, 'Invalid parameters count');
    assertErrors(context);
});
test('loadEmptyMap', () => {
    let chain = readChain('[:]');
    let context = new Context(chain.get());
    let call = new CallArray(context, chain.get());
    call.load();
    assertEquals(0, call.getMembers().length, 'Invalid parameters count');
    assertErrors(context);
});
test('loadEmptyGrid', () => {
    let chain = readChain('[[]]');
    let context = new Context(chain.get());
    let call = new CallArray(context, chain.get());
    call.load();
    assertEquals(0, call.getMembers().length, 'Invalid parameters count');
    assertErrors(context);
});
test('loadMixTypesArrayGrid', () => {
    let chain = readChain('[1, [3, 4]]');
    let context = new Context(chain.get());
    let call = new CallArray(context, chain.get());
    call.load();
    assertEquals(2, call.getMembers().length, 'Invalid parameters count');
    assertErrors(context, Error.arrayMixingTypes);
});
test('loadMixTypesArrayMap', () => {
    let chain = readChain('[1, a : b]');
    let context = new Context(chain.get());
    let call = new CallArray(context, chain.get());
    call.load();
    assertEquals(2, call.getMembers().length, 'Invalid parameters count');
    assertErrors(context, Error.arrayMixingTypes);
});
test('loadMixTypesGridMap', () => {
    let chain = readChain('[[a], a : b]');
    let context = new Context(chain.get());
    let call = new CallArray(context, chain.get());
    call.load();
    assertEquals(2, call.getMembers().length, 'Invalid parameters count');
    assertErrors(context, Error.arrayMixingTypes);
});
test('loadDoubleComma_Fail', () => {
    let chain = readChain('[a,,b]');
    let context = new Context(chain.get());
    let call = new CallArray(context, chain.get());
    call.load();
    assertEquals(2, call.getMembers().length, 'Invalid parameters count');
    assertErrors(context, Error.unexpectedToken);
});
test('loadSingleComma_Fail', () => {
    let chain = readChain('[,]');
    let context = new Context(chain.get());
    let call = new CallArray(context, chain.get());
    call.load();
    assertEquals(0, call.getMembers().length, 'Invalid parameters count');
    assertErrors(context, Error.unexpectedToken);
});
test('loadEndComma_Fail', () => {
    let chain = readChain('[a,]');
    let context = new Context(chain.get());
    let call = new CallArray(context, chain.get());
    call.load();
    assertEquals(1, call.getMembers().length, 'Invalid parameters count');
    assertErrors(context, Error.unexpectedEndOfTokens);
});
test('loadCommaBetweenTernary_Fail', () => {
    let chain = readChain('[a ? b , : c]');
    let context = new Context(chain.get());
    let call = new CallArray(context, chain.get());
    call.load();
    assertEquals(2, call.getMembers().length, 'Invalid parameters count');
    assertErrors(context, Error.lineTernaryIncomplete, Error.lineTernaryIncomplete);
});
test('loadMapArray_Fail', () => {
    let chain = readChain('[a : [1]]');
    let context = new Context(chain.get());
    let call = new CallArray(context, chain.get());
    call.load();
    assertEquals(1, call.getMembers().length, 'Invalid parameters count');
    assertErrors(context, Error.arrayContainer);
});
test('loadMapKeyArray_Fail', () => {
    let chain = readChain('[[a] : 1]');
    let context = new Context(chain.get());
    let call = new CallArray(context, chain.get());
    call.load();
    assertEquals(1, call.getMembers().length, 'Invalid parameters count');
    assertErrors(context, Error.arrayContainer);
});
test('loadGridArray_Fail', () => {
    let chain = readChain('[[[a]]]');
    let context = new Context(chain.get());
    let call = new CallArray(context, chain.get());
    call.load();
    assertEquals(1, call.getMembers().length, 'Invalid parameters count');
    assertErrors(context, Error.arrayContainer);
});
test('loadMissingCloser_Fail', () => {
    let chain = readChain('[1');
    let context = new Context(chain.get());
    let call = new CallArray(context, chain.get());
    call.load();
    assertEquals(1, call.getMembers().length, 'Invalid parameters count');
    assertErrors(context, Error.missingCloser);
});
test('loadMissingInnerCloser_Fail', () => {
    let chain = readChain('[[1');
    let context = new Context(chain.get());
    let call = new CallArray(context, chain.get());
    call.load();
    assertEquals(1, call.getMembers().length, 'Invalid parameters count');
    assertErrors(context, Error.missingCloser, Error.missingCloser);
});
test('loadInnerComma_Fail', () => {
    let chain = readChain('[[1,]]');
    let context = new Context(chain.get());
    let call = new CallArray(context, chain.get());
    call.load();
    assertEquals(1, call.getMembers().length, 'Invalid parameters count');
    assertErrors(context, Error.unexpectedEndOfTokens);
});
test('loadInnerDoubleComma_Fail', () => {
    let chain = readChain('[[a,,b]]');
    let context = new Context(chain.get());
    let call = new CallArray(context, chain.get());
    call.load();
    assertEquals(1, call.getMembers().length, 'Invalid parameters count');
    assertErrors(context, Error.unexpectedToken);
});
